File: Java/src/com/venus/List/CycleLinkedList.java
package com.venus.List;

/**
 * @copyright 熊彪
 *  * 单向循环链表
 *  * 注意点：除链表注意点外，循环链表添加删除元素需对头部进行特殊处理，将尾部next指向新的头部。删除节点如果只有一个元素将head变成null
 *  * 优点：节省空间
 *  * 缺点：可能造成内存空间浪费
 *   add :    O(1) O(n)
 *   remove : O(1) O(n)
 *   get :    O(1) O(n)
 *   set :    O(1) O(n)
 */

@SuppressWarnings("all")
public class CycleLinkedList<E> extends AbstractList<E>{
    private Node head;

    private static class Node<E> {
        E element;
        Node<E> next;

        public Node(E element, Node<E> next) {
            super();
            this.element = element;
            this.next = next;
        }
    }


    @Override
    public void clear() {
        size = 0;
        head = null;
    }

    @Override
    public void add(E element, int index) {
        rangeCheckForAdd(index);
        if (index == 0){
            Node<E> newNode = new Node(element, head);
            Node<E> last = (size == 0) ? newNode : node(size-1);
            last.next = newNode;
            head = newNode;
        }else {
            Node prev = node(index - 1);
            prev.next = new Node(element, prev.next);
        }
        size++;
    }

    private Node<E> node(int index) {
        rangeCheck(index);
        Node node = head;
        for (int i = 0; i < index; i++) {
            node = node.next;
        }
        return node;
    }

    @Override
    public E get(int index) {
        return node(index).element;
    }

    @Override
    public E set(int index, E element) {
        Node<E> node = node(index);
        E oldElement = node.element;
        node.element = element;
        return oldElement;
    }

    @Override
    public E remove(int index) {
        rangeCheck(index);
        Node<E> removeElement = head;
        if (index == 0){
            if (size == 1){
                head = null;
            }else {
                head = head.next;
                Node<E> last = node(size - 1);
                last.next = head;
            }
        }else {
            Node<E> prev = node(index - 1);
            removeElement = prev.next;
            prev.next = removeElement.next;
        }
        size--;
        return removeElement.element;
    }

    @Override
    public int indexOf(E element) {
        Node<E> node = head;
        if (element == null){
            for (int i = 0; i < size; i++) {
                if (element == node.element) return i;
                node = node.next;
            }
        }else {
            for (int i = 0; i < size; i++) {
                if (element.equals(node.element)) return i;
                node = node.next;
            }
        }
        return -1;
    }


    @Override
    public String toString() {
        StringBuilder s = new StringBuilder();
        s.append("Size: ").append(size).append(", LinkedList: [");
        Node node = head;
        for (int i = 0; i < size; i++) {
            if (i != 0) s.append(", ");
            s.append(node.element);
            node = node.next;
        }
        s.append("]");
        return s.toString();
    }
}

File: Python/List/cyclelinkedlist.py

## written by xiongbiao
## date 2020-5-19


from List.list import List
from List.linkedlist import Node

class CycleLinkedList(List):
    def __init__(self, nums=None):
        super().__init__()
        self.__head = None
        if nums:
            for n in nums:
                self.add_value(n)
    '''
    添加节点，空链表，插入头结点，其他节点
    '''
    def add_value_at(self, value, index):
        self._range_check_for_add(index)
        if index == 0:
            last = self._get_node(self._size - 1)

            new_node = Node(value, self.__head)
            if last is None:
                new_node.next = new_node
            else:
                last.next = new_node
            self.__head = new_node
        else:
            node = self._get_node(index - 1)
            new_node = Node(value, node.next)
            node.next = new_node

        self._size += 1
    '''
    删除某个位置节点，删除头结点，当节点只有一个元素和节点有多于一个元素特殊处理；删除其他节点
    '''
    def remove_value_at(self, index):
        self._range_check(index)
        if index == 0:
            value = self.__head.value
            if self._size == 1:
                self.__head = None
            else:
                last = self._get_node(self._size - 1)
                self.__head = self.__head.next
                last.next = self.__head
        else:
            prev = self._get_node(index - 1)
            value = prev.next.value
            prev.next = prev.next.next

        self._size -= 1

        return value

    def _get_node(self, index):
        node = self.__head
        for i in range(index):
            node = node.next

        return node

    '''
        获取index的value
        '''

    def get_value(self, index):
        self._range_check(index)
        node = self._get_node(index)
        return node.value

    '''
    修改index的value
    '''

    def set_value(self, index, value):
        self._range_check(index)

        node = self._get_node(index)
        old_value = node.value
        node.value = value
        return old_value

    def __repr__(self):
        s = 'LinkedList=['
        node = self.__head

        for i in range(0, self._size):
            if i == 0:
                s += '{}(next: {})'.format(node.value, node.next.value)
            else:
                s += ', {}(next: {})'.format(node.value, node.next.value)
            node = node.next
        s += '], Size: {}'.format(self._size)

        return s


cycle = CycleLinkedList([1,2,3,4,5,6,7,8])
# cycle.add_value_at(10, 0)
# cycle.add_value_at(15, 9)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
cycle.remove_value_at(0)
# cycle.remove_value_at(0)

print(cycle)